examples/text_to_image/train_text_to_image.py

```diff
@@ -396,6 +408,31 @@ def main():
         else:
             raise ValueError("xformers is not available. Make sure it is installed correctly")
 
+    # `accelerate` 0.15.0 will have better support for customized saving
+    if version.parse(accelerate.__version__) >= version.parse("0.15.0.dev0"):
+        # create custom saving & loading hooks so that `accelerator.save_state(...)` serializes in a nice format
+        def save_model_hook(models, weights, output_dir):
+            for i, model in enumerate(models):
+                model.save_pretrained(os.path.join(output_dir, "unet"))
+
+                # make sure to pop weight so that corresponding model is not saved again
+                weights.pop()
+
+        def load_model_hook(models, input_dir):
+            for i in range(len(models)):
+                # pop models so that they are not loaded again
+                model = models.pop()
+
+                # load diffusers style into model
+                load_model = UNet2DConditionModel.from_pretrained(input_dir, subfolder="unet")
+                model.register_to_config(**load_model.config)
+
+                model.load_state_dict(load_model.state_dict())
+                del load_model
+
+        accelerator.register_save_state_pre_hook(save_model_hook)
```

Comment:
Here is it also possible to extend this for `ema_unet` ? It's registered for checkpointing using `register_for_checkpointing` and will be saved as `custom_checkpoints0.bin`. For text_to_image, we need to load the ema weights for inference.

Reply:
That's a very good point. I tested both `train_text_to_image` and `train_unconditional` and I saw that:

- The ema model is not sent to the `save_model_hook` function despite being registered for checkpointing. I suppose the hook only deals with weights being trained.
- In `train_unconditional` no `custom_checkpoints0.bin` is saved. I don't know the reason why.

If we need to save ema weights then we might need to save the model ourselves.

Reply:
I think It's not passed to `save_model_hook` because it's not registered as a `nn.Moudle` in with `accelerate.prepare`. We could also, instead make the `EMAModel` an instance of `nn.Module` and pass to `prepare` so it'll work with `save_model_hook`

Reply:
Oh, you are right! Do you think it's better to make it a `nn.Module` or handle it as a special case in the training scripts?

Reply:
Nice catch @patil-suraj ! 

I would be in favor of making the `EMAModel` a `torch.nn.Module` so that it's recognized automatically by `accelerate` here: https://github.com/huggingface/accelerate/blob/d0df263b0917773935110411a347cc3628213342/src/accelerate/checkpointing.py#L138

The EMA model class defines a very nice save & load state dict functions here: https://github.com/huggingface/diffusers/blob/87cf88ed3d351cc4e2b7cdd462ddf7a4ebf2109e/src/diffusers/training_utils.py#L201 which include all the important parameters.

Then I'd slightly need to adapt the hooks to not pop the EMAModel and let accelerate handle the saving and loading. 

Think that's a nice solution because:
- EMA model has to config
- IMO it's not strictly necessary to have a nice folder structure for the EMA model during training as the "non-ema" UNet should be checked for training progress instead. I'd argue, it's easier to directly see with the non-EMA model whether training is overfitting or not? 

Wdyt?

Reply:
Yeah I agree, no harm in making `EMAModel` an instance of `nn.Module`. 


>I'd argue, it's easier to directly see with the non-EMA model whether training is overfitting or not?

That's true, but if we are doing generation during training, ideally we would want to use the ema model, as that's the final model we use for actual inference.